### PR TITLE
Mtest annotations and bug fixes

### PR DESCRIPTION
**Pull request by dcbaker**
This started out as me thinking "hey mypy can help me fix [this](https://gitlab.freedesktop.org/idr/mesa/-/jobs/298246)!" and ended with me cursing loudly. This code was pretty difficult to follow and had some pretty terrible things (changing the types of attributes silently). I think mtest is a little cleaner now, and the annotations helped my find both the bug I was looking for and a few others that I didn't know where there. So if you've ever questioned the value of type annotations, they're paying off here.

### CODE REVIEW

File: mesonbuild/mtest.py [hunk 1]
```diff
@@ -346,20 +364,23 @@ def make_tap(test, returncode, duration, stdo, stde, cmd):
...
                 res = TestResult.FAIL if failed else TestResult.OK
 
-        return TestRun(test, res, returncode, duration, stdo, stde, cmd)
+        return cls(test, test_env, res, returncode, duration, stdo, stde, cmd)
 
-    def __init__(self, test, res, returncode, duration, stdo, stde, cmd):
+    def __init__(self, test: 'TestSerialisation', test_env: typing.Dict[str, str],
+                res: TestResult, returncode: int, duration: float,
+                stdo: typing.Optional[str], stde: typing.Optional[str],
+                cmd: typing.Optional[typing.List[str]]):
```
Comment (ignatenkobrain) [inline mesonbuild/mtest.py]: __[Flake8]__ [E128] continuation line under-indented for visual indent ([link](https://sider.review/gh/repos/19784232/pulls/5395#issue-4079294))

<sub>posted by Sider</sub>

File: mesonbuild/mtest.py [hunk 2]
```diff
@@ -346,20 +364,23 @@ def make_tap(test, returncode, duration, stdo, stde, cmd):
             else:
                 res = TestResult.FAIL if failed else TestResult.OK
 
-        return TestRun(test, res, returncode, duration, stdo, stde, cmd)
+        return cls(test, test_env, res, returncode, duration, stdo, stde, cmd)
 
-    def __init__(self, test, res, returncode, duration, stdo, stde, cmd):
+    def __init__(self, test: 'TestSerialisation', test_env: typing.Dict[str, str],
+                res: TestResult, returncode: int, duration: float,
```
Comment (ignatenkobrain) [inline mesonbuild/mtest.py]: __[Flake8]__ [E128] continuation line under-indented for visual indent ([link](https://sider.review/gh/repos/19784232/pulls/5395#issue-4079292))

<sub>posted by Sider</sub>

File: mesonbuild/mtest.py [hunk 3]
```diff
@@ -346,20 +364,23 @@ def make_tap(test, returncode, duration, stdo, stde, cmd):
...
+                stdo: typing.Optional[str], stde: typing.Optional[str],
```
Comment (ignatenkobrain) [inline mesonbuild/mtest.py]: __[Flake8]__ [E128] continuation line under-indented for visual indent ([link](https://sider.review/gh/repos/19784232/pulls/5395#issue-4079293))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
I'll fix the sider errors tomorrow.

File: mesonbuild/build.py [hunk 4]
```diff
@@ -140,7 +141,7 @@ def __init__(self, environment):
         self.dep_manifest_name = None
         self.dep_manifest = {}
         self.cross_stdlibs = {}
-        self.test_setups = {}
+        self.test_setups = {}                         # type: typing.Dict[str, TestSetup]
```
Comment (Ericson2314) [inline mesonbuild/build.py]: Big indent here on purpose?

Comment (dcbaker) [reply mesonbuild/build.py]: Yeah, I wanted to be able to line all of the type annotations for those attributes up like:

```python
self.foo = []                     # type: ...
self.really_long_name = []        # type: ...
```

**Pull request merged**